pkg/reconciler/credentials/reconciler.go

```diff
@@ -0,0 +1,73 @@
+/*
+Copyright 2019 kubeflow.org.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package credentials
+
+import (
+	"context"
+	knservingv1alpha1 "github.com/knative/serving/pkg/apis/serving/v1alpha1"
+	"github.com/kubeflow/kfserving/pkg/constants"
+	"github.com/kubeflow/kfserving/pkg/reconciler/credentials/gcs"
+	"github.com/kubeflow/kfserving/pkg/reconciler/credentials/s3"
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+)
+
+var log = logf.Log.WithName("CredentialReconciler")
+
+type CredentialReconciler struct {
+	client client.Client
+}
+
+func NewCredentialReconciler(client client.Client) *CredentialReconciler {
+	return &CredentialReconciler{
+		client: client,
+	}
+}
+
+func (c *CredentialReconciler) ReconcileServiceAccount(ctx context.Context, namespace string, serviceAccountName string,
```

Comment:
Not crazy about "reconciling" here. We're not actually reconciling resources (i.e., observing & enforcing the state of the world), we're mutating a configuration which needs to be enforced elsewhere. The way I see it, we only want one reconciler per sub-resource manage (i.e., Route/Configuration). 

I think this might be cleaner if we have a pattern (in the caller) like:
```
envVars = BuildCredentialEnvVars(modelSpec.serviceAccountName)
configuration.Spec.RevisionTemplate.Spec.Container.Env = append(configuration.Spec.RevisionTemplate.Spec.Container.Env, envs...)
```

Typically, I feel that mutation, especially in increasingly vertical code bases, can get really difficult to manage. It's always hard to reason about what side effects are going to happen to your resource when you call a function.

Now if only we could get a functional programming language that's built for the declarative infra world ;)

WDYT?

Reply:
I actually implemented the way you suggested initially, was thinking if needs to make the mutation more explicit. I think we are also reconciling secrets here since that can also change, maybe needs better wording here?

Reply:
I'm actually waffling between the two ways -- as I guess you are too. In some ways mutation is cleaner as you can delegate the assignment and just treat the top level as a workflow executor. 

I'd have a better idea of how to fix it if I were neck deep in the code. I'll trust your judgement here, but if it continues to bother me, I might take a stab at a refactor ;)

Reply:
I also feel reconclile doesn't fit. You say reconcile will fire when a secret changes. I thought you needed to setup a `Watch` for this which I don't see. Is there a test for this?

Reply:
There is no ownership between secret and kfservice so not possible to setup a watch. But there is controller resync which will move the state. I agree it does not sound reconcile and I am open for suggestions.  I already refactored the mutation code back to stateless.

Reply:
What about https://godoc.org/sigs.k8s.io/controller-runtime/pkg#hdr-Watching_and_EventHandling

```EnqueueRequestFromMapFunc```

This should probably live at the configuration level and drive changes to the configuration. I wouldn't worry to much about the name since no one has an adequate suggestion, but we may want to look into a refactor at some point.

Reply:
SGTM, I am trying to restructure the files a bit.